Allow setting policies on actions

chmouel opened the pull request:
Policies is a new concept which let you allowing teams to do some
actions and reject member who are not part of those teams.

Current actions supported are `pull_request` and `ok_to_test` (more to
come in the future)

See the documentation attached to this PullRequest for more description
on how to use this feature.

chmouel committed:
Refactor filtering of the eventype to its own func

so we can reuse it later on

chmouel commented:
weird, things are green for me locally when running the test locally

```
% e2e-run -k kind -p gitea-kind -n TestGiteaACL
Running TestGiteaACL
Cleaning test cache
=== RUN   TestGiteaACLOrgAllowed
💡 06:20:51 Namespace pac-e2e-test-vb9sq created
💡 06:20:51 Creating gitea repository pac-e2e-test-vb9sq for user pac
💡 06:20:52 Creating webhook to smee url on gitea repository pac-e2e-test-vb9sq
💡 06:20:52 PipelinesAsCode Repository pac-e2e-test-vb9sq has been created in namespace pac-e2e-test-vb9sq
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:20:53 Pushed files to repo http://localhost:3000/pac/pac-e2e-test-vb9sq branch pac-e2e-test-vb9sq
💡 06:20:54 PullRequest http://localhost:3000/pac/pac-e2e-test-vb9sq/pulls/1 has been created
💡 06:20:55 Forked repository http://localhost:3000/pac-e2e-test-vb9sq/pac-e2e-test-vb9sq.git has been created
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:20:56 Pushed files to repo  branch pac-e2e-test-vb9sq
💡 06:21:00 Created pr http://localhost:3000/pac/pac-e2e-test-vb9sq/pulls/2 branch:main from fork pac-e2e-test-vb9sq/pac-e2e-test-vb9sq, branch:pac-e2e-test-vb9sq
💡 06:21:17 Status on SHA: pac-e2e-test-vb9sq is success
--- PASS: TestGiteaACLOrgAllowed (25.79s)
=== RUN   TestGiteaACLOrgSkipped
💡 06:21:17 Namespace pac-e2e-test-tt8z4 created
💡 06:21:17 Creating gitea repository pac-e2e-test-tt8z4 for user pac
💡 06:21:18 Creating webhook to smee url on gitea repository pac-e2e-test-tt8z4
💡 06:21:18 PipelinesAsCode Repository pac-e2e-test-tt8z4 has been created in namespace pac-e2e-test-tt8z4
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:19 Pushed files to repo http://localhost:3000/pac/pac-e2e-test-tt8z4 branch pac-e2e-test-tt8z4
💡 06:21:26 PullRequest http://localhost:3000/pac/pac-e2e-test-tt8z4/pulls/1 has been created
💡 06:21:34 Forked repository http://localhost:3000/pac-e2e-test-tt8z4/pac-e2e-test-tt8z4.git has been created
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:35 Pushed files to repo  branch pac-e2e-test-tt8z4
💡 06:21:38 Created pr http://localhost:3000/pac/pac-e2e-test-tt8z4/pulls/2 branch:main from fork pac-e2e-test-tt8z4/pac-e2e-test-tt8z4, branch:pac-e2e-test-tt8z4
💡 06:21:48 Status on SHA: 73f9452288ea578d13e1cbce74a2a6aa6144ec5e is success
💡 06:21:49 Found regexp ".*is skipping this commit.*" in PR comments
--- PASS: TestGiteaACLOrgSkipped (31.68s)
=== RUN   TestGiteaACLCommentsAllowing
=== RUN   TestGiteaACLCommentsAllowing/OK_to_Test
💡 06:21:49 Namespace pac-e2e-test-hx7rv created
💡 06:21:49 Creating gitea repository pac-e2e-test-hx7rv for user pac
💡 06:21:49 Creating webhook to smee url on gitea repository pac-e2e-test-hx7rv
💡 06:21:49 PipelinesAsCode Repository pac-e2e-test-hx7rv has been created in namespace pac-e2e-test-hx7rv
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:50 Pushed files to repo http://localhost:3000/pac/pac-e2e-test-hx7rv branch pac-e2e-test-hx7rv
💡 06:21:50 PullRequest http://localhost:3000/pac/pac-e2e-test-hx7rv/pulls/1 has been created
💡 06:21:51 Forked repository http://localhost:3000/pac-e2e-test-hx7rv/pac-e2e-test-hx7rv.git has been created
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:51 Pushed files to repo  branch pac-e2e-test-hx7rv
💡 06:21:52 Created pr http://localhost:3000/pac/pac-e2e-test-hx7rv/pulls/2 branch:main from fork pac-e2e-test-hx7rv/pac-e2e-test-hx7rv, branch:pac-e2e-test-hx7rv
💡 06:21:57 Status on SHA: c7c80483769aa088b9cc31f2631ba90f6117660b is success
💡 06:21:57 Found regexp ".*is skipping this commit.*" in PR comments
💡 06:21:57 Posted comment "/ok-to-test" in http://localhost:3000/pac/pac-e2e-test-hx7rv/pulls/2
💡 06:21:57 Found regexp "^Pipelines as Code CI.*has.*successfully" in PR comments
=== RUN   TestGiteaACLCommentsAllowing/Retest
💡 06:21:57 Namespace pac-e2e-test-xl4r4 created
💡 06:21:57 Creating gitea repository pac-e2e-test-xl4r4 for user pac
💡 06:21:57 Creating webhook to smee url on gitea repository pac-e2e-test-xl4r4
💡 06:21:57 PipelinesAsCode Repository pac-e2e-test-xl4r4 has been created in namespace pac-e2e-test-xl4r4
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:58 Pushed files to repo http://localhost:3000/pac/pac-e2e-test-xl4r4 branch pac-e2e-test-xl4r4
💡 06:21:58 PullRequest http://localhost:3000/pac/pac-e2e-test-xl4r4/pulls/1 has been created
💡 06:21:59 Forked repository http://localhost:3000/pac-e2e-test-xl4r4/pac-e2e-test-xl4r4.git has been created
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:21:59 Pushed files to repo  branch pac-e2e-test-xl4r4
💡 06:22:00 Created pr http://localhost:3000/pac/pac-e2e-test-xl4r4/pulls/2 branch:main from fork pac-e2e-test-xl4r4/pac-e2e-test-xl4r4, branch:pac-e2e-test-xl4r4
💡 06:22:05 Status on SHA: d5e741bb6b2d935008a0937528428e19656599e4 is success
💡 06:22:05 Found regexp ".*is skipping this commit.*" in PR comments
💡 06:22:05 Posted comment "/retest" in http://localhost:3000/pac/pac-e2e-test-xl4r4/pulls/2
💡 06:22:05 Found regexp "^Pipelines as Code CI.*has.*successfully" in PR comments
=== RUN   TestGiteaACLCommentsAllowing/Test_PR
💡 06:22:05 Namespace pac-e2e-test-8sxlq created
💡 06:22:05 Creating gitea repository pac-e2e-test-8sxlq for user pac
💡 06:22:05 Creating webhook to smee url on gitea repository pac-e2e-test-8sxlq
💡 06:22:06 PipelinesAsCode Repository pac-e2e-test-8sxlq has been created in namespace pac-e2e-test-8sxlq
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:22:06 Pushed files to repo http://localhost:3000/pac/pac-e2e-test-8sxlq branch pac-e2e-test-8sxlq
💡 06:22:06 PullRequest http://localhost:3000/pac/pac-e2e-test-8sxlq/pulls/1 has been created
💡 06:22:07 Forked repository http://localhost:3000/pac-e2e-test-8sxlq/pac-e2e-test-8sxlq.git has been created
    scm.go:44: skipping cleanup because TEST_NOCLEANUP was enabled.
    scm.go:50: skipping cleanup because TEST_NOCLEANUP was enabled.
💡 06:22:07 Pushed files to repo  branch pac-e2e-test-8sxlq
💡 06:22:08 Created pr http://localhost:3000/pac/pac-e2e-test-8sxlq/pulls/2 branch:main from fork pac-e2e-test-8sxlq/pac-e2e-test-8sxlq, branch:pac-e2e-test-8sxlq
💡 06:22:13 Status on SHA: 4a54977e031c3b27219795491b73fa185cf4c07a is success
💡 06:22:13 Found regexp ".*is skipping this commit.*" in PR comments
💡 06:22:13 Posted comment "/test pr" in http://localhost:3000/pac/pac-e2e-test-8sxlq/pulls/2
💡 06:22:13 Found regexp "^Pipelines as Code CI.*has.*successfully" in PR comments
--- PASS: TestGiteaACLCommentsAllowing (24.37s)
    --- PASS: TestGiteaACLCommentsAllowing/OK_to_Test (8.34s)
    --- PASS: TestGiteaACLCommentsAllowing/Retest (8.05s)
    --- PASS: TestGiteaACLCommentsAllowing/Test_PR (7.98s)
PASS
ok  	github.com/openshift-pipelines/pipelines-as-code/test	81.868s
```

chmouel commented:
nevermind the other comment the error was legitimate

sm43 commented:
what does `pull_request` action means here? 🤔

chmouel commented:
@sm43 it's this


```markdown
* `pull_request` - This action is triggering the CI on Pipelines as Code,
   specifying a team will only allow the members of the team to trigger the CI
   and will not allow other members regadless if they are Owners or Collaborators
   of the repository or the Organization. The OWNERS file is still taken into
   account and will as well allow the members of the OWNERS file to trigger the
   CI.
```

sm43 commented on pkg/provider/github/acl.go:
```diff
@@ -9,9 +9,55 @@ import (
+	"github.com/openshift-pipelines/pipelines-as-code/pkg/policy"
 )
 
+// CheckPolicyAllowing check that policy is
```
comment is incomplete?

sm43 reviewed (approved):
lgtm

chmouel committed:
Fix a bunch of gitea e2e random failures

with regard to the latest change setting pipelinerunpending before
running it.

we need to make sure we only get the latest status to properly test

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Make sure we test for Pending Approval

When the user is not allowed make sure we check the test are in Pending
Approval

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

savitaashture commented on test/gitea_access_control_test.go:
```diff
@@ -198,7 +198,7 @@ func TestGiteaACLOrgPendingApproval(t *testing.T) {
 
 	topts.PullRequest = tgitea.CreateForkPullRequest(t, topts, secondcnx, "", "echo Hello from user "+topts.TargetRefName)
-	topts.CheckForStatus = "pending"
+	topts.CheckForStatus = "Skipped"
```
why skipped :thinking:

chmouel replied:
because it was ?

savitaashture commented on docs/content/docs/guide/policy.md:
```diff
@@ -0,0 +1,63 @@
+title: Policy on actions
+weight: 9
+---
+# Policy on Pipeline as Code actions
```
```suggestion
# Policy on Pipelines as Code actions
```

savitaashture commented on config/300-repositories.yaml:
```diff
@@ -70,6 +70,20 @@ spec:
+                        pull_request:
+                          type: array
+                          items:
+                            description: list of teams allowed to have ci run on pull/merge requests.
```
I see feature is supported only for GitHub/Gitea does merge will be applicable here
I see merge is used for GitLab :thinking:

chmouel replied:
yeah, in the future we will add more providers...

savitaashture commented on docs/content/docs/guide/policy.md:
```diff
@@ -0,0 +1,63 @@
+   specifying a team will only allow the members of the team to trigger the CI
+   and will not allow other members regadless if they are Owners or Collaborators
+   of the repository or the Organization. The OWNERS file is still taken into
+   account and will as well allow the members of the OWNERS file to trigger the
```
Even though members of the OWNER files not belongs to that particular team

savitaashture commented on docs/content/docs/guide/policy.md:
````diff
@@ -0,0 +1,63 @@
+        - ci-users
+```
+
+Users in `ci-admins` team will be able to let other users run the CI on the pull
````
I feel explanation is somewhat confusing to tell the exact operation which `ci-admins` can do 
May its for me :thinking:

chmouel replied:
is that less confusing ? (chatgpt)

To set up the policy for the Repository CR, follow these steps:

1. Open the settings of the Repository CR.
2. Add the following code snippet to the configuration:

```yaml
apiVersion: "pipelinesascode.tekton.dev/v1alpha1"
kind: Repository
metadata:
  name: repository1
spec:
  url: "https://github.com/org/repo"
  settings:
    policy:
      ok_to_test: 
        - ci-admins
      pull_request: 
        - ci-users
```

This configuration allows specific user groups to perform certain actions on pull requests:

- Users in the `ci-admins` team can authorize other users to run continuous integration (CI) on the pull request.
- Users in the `ci-users` team can run CI on their own pull requests.

savitaashture replied:
@chmouel this is easy to understand :+1:

savitaashture commented on pkg/policy/policy.go:
```diff
@@ -0,0 +1,60 @@
+	var sType []string
+	switch tType {
+	// NOTE: This make /retest /ok-to-test /test bound to the same policy, which is fine from a security standpoint but maybe we want to refind
+	case info.TriggerTypeOkToTest, info.TriggerTypeRetest:
```
I see case handles for `/retest, /ok-to-test` not handle for `/test` as per the comment

savitaashture commented on pkg/params/info/events.go:
```diff
@@ -89,3 +91,15 @@ func NewEvent() *Event {
+
+const (
+	TriggerTypeOkToTest              TriggerType = "ok-to-test"
+	TriggerTypeRetest                TriggerType = "retest"
```
I think need to add const for `/test` as well :thinking:

chmouel replied:
/test and /retest are the same in the code

chmouel replied:
see here https://github.com/openshift-pipelines/pipelines-as-code/blob/2ced6ceb7f44ab285bebf52ed2f624c5e5b23663/pkg/provider/provider.go#L37

savitaashture commented on pkg/policy/policy.go:
```diff
@@ -0,0 +1,60 @@
+	case info.TriggerTypePullRequest:
+		sType = p.Settings.Policy.PullRequest
+	// NOTE: not supported yet, will imp if it gets requested and reasonable to implement
+	case info.TriggerTypePush, info.TriggerTypeCancel, info.TriggerTypeCheckSuiteRerequested, info.TriggerTypeCheckRunRerequested:
+		return ResultNotSet, nil
+	default:
+		return ResultNotSet, nil
+	}
```
Can we skip the case and for now add a comment and return `ResultNotSet` from default case

chmouel replied:
i left them here as kinda todo list for the future to implement... will make it easier when we get there...

savitaashture commented on test/gitea_access_control_test.go:
```diff
@@ -0,0 +1,310 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TestGiteaPolicyOkToTestRetest test the policy on pull request scenario
```
```suggestion
// TestGiteaPolicyPullRequest test the policy on pull request scenario
```

chmouel committed:
Allow setting policies on actions

Policies is a new concept which let you allowing teams to do some
actions and reject member who are not part of those teams.

Current actions supported are `pull_request` and `ok_to_test` (more to
come in the future)

See the documentation attached to this PullRequest for more description
on how to use this feature.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
/retest

chmouel committed:
Merge branch 'main' into policies-setting

savitaashture commented:
/lgtm

piyush-garg commented:
/lgtm
